### ISSUE TITLE
Versioning convention + show last commit?

### ISSUE THREAD

**Issue by dexX7** [mastercoin-MSC/mastercore#155]
Release v0.0.8.1 has a `mastercoreversion` of `11008` and furthermore dev-versions come with a prefixed value, e.g. the dev-versions between 0.0.6 and 0.0.7 used at some point `1310006` which later became the tagged release v0.0.7 with the value `10007`.

Since the `mastercoreversion` field is used by humans (I guess?), I suggest to use a more readable format - for example "0.0.6.31" or to be even more explicit: "0.0.6.31-dev". Adding a new field which either shows "release" or "development" may also be an option.

---

I have several boxes and I'm sort of unorganized, and sometimes I end up with different builds on different boxes asking myself: "which build is actually running right now?"

Since bumping version numbers is a rather laborious task and there is no strong relation between version numbers, commits and builds to begin with, I was wondering: is there a way to add a step to the build process to make the binaries self-aware such that they are able to show the last commit that was used for a build?

P.S.: good read: http://semver.org/

**Comment by dexX7** [comment]
I noticed this is partially done and might be extended. When building Master or Bitcoin Core, the file [/share/genbuild.sh](https://github.com/mastercoin-MSC/mastercore/blob/mscore-0.0.9/share/genbuild.sh) is called. It creates the file /obj/build.h which contains the following information:

```
#define BUILD_SUFFIX 9de5ee0-dirty
#define BUILD_DATE "2014-12-06 22:36:58 +0100"
```

In the case it's no tagged release, a partial commit hash is written and otherwise the release version. This information is picked up on the GUI start and shown on the splash screen (not sure, if this is currently overwritten).

Shouldn't be too difficult to extend or to be accessed. @zathras-crypto: this might be handy for the alert system and more generally, this may replace hardcoded version strings in the RPC output.

Edit: this information is available via [clientversion.h](https://github.com/mastercoin-MSC/mastercore/blob/mscore-0.0.9/src/clientversion.h#L5).

**Comment by zathras-crypto** [comment]
Thanks DexX :)

OK perhaps I should spend a few moments just talking about the changes recently.  So versioning has been substantially changed along with the alert PR.  There is no longer a 5 digit int that we just display in a hacked rpcmisc.cpp where we have no Master Protocol support (ie getinfo call for bitcoin).  

Instead there are now 2 fields and these are defined in mastercore.h so we have this versioning within the master side of 'core' (and thus usable for alerts, version details now in getinfo_MP call):

```
+// define the version for alert messages
+#define MASTERCORE_VERSION_BASE 90 // 82 = 0.0.8.2   91 = 0.0.9.1   103 = 0.0.10.3 etc
+#define MASTERCORE_VERSION_TYPE "-dev" // switch to -rel for tags, switch back to -dev for development
```

The version supplied in getinfo_MP is worked out as follows:
1) String "0.0."
2) Append string of (MASTERCORE_VERSION_BASE / 10) (eg 90 = 9, 91 = 9.1 etc)
3) Append MASTERCORE_VERSION_TYPE
That gives something along the lines of "0.0.9-dev" or "0.0.9.1-rel" from the example above.
Specifically:

```
// provide the mastercore and bitcoin version
infoResponse.push_back(Pair("mastercoreversion", "0.0." + boost::lexical_cast<string>((double)MASTERCORE_VERSION_BASE/10) + MASTERCORE_VERSION_TYPE ));
infoResponse.push_back(Pair("bitcoincoreversion", "0." + boost::lexical_cast<string>((int)CLIENT_VERSION/100)));
```

The idea being that for each build we release (not necessarily each build we _do_) the BASE version should be incremented by one and the -dev tag should be flipped to -rel for a release and back to -dev again aftwerwards.  FYI the BASE version int is the value alerts expired on client version will compare against.  Thus for now to keep things simple, I want this to be a simple int while alerts see their first iteration just to make sure comparisons of "is client under X version" are simple at first.

So, in a nutshell - we should (from now on) already be incrementing versions and denoting dev or release builds (-dev/-rel) in the version.  

I hope that helps to clarify what I've changed.  With all that said though, I love the idea of having the commit available as well (especially people doing their own builds for example or running -dev versions), so have added that mate - thanks for suggestion.

FYI getinfo_MP output example for tip is here (commit field is to be merged but rest is already in tip): http://pastie.org/pastes/9768658/text?key=tg0lmgx5wwebb7idrrnr8q

Thanks :)
Z

**Comment by dexX7** [comment]
> The idea being that for each build we release (not necessarily each build we do) the BASE version should be incremented by one and the -dev tag should be flipped to -rel for a release and back to -dev again aftwerwards.

Assuming releases are for the broader public, isn't appending -rel sort of just stating the obvious?

Regarding the versioning scheme, I like the idea. I still recommend a look at http://semver.org. The summary is already down to the point and one of the very often used approaches on the web:

> Given a version number **MAJOR.MINOR.PATCH**, increment the:
> 
> **MAJOR** version when you make **incompatible** API changes,
> **MINOR** version when you add functionality in a **backwards-compatible** manner, and
> **PATCH** version when you make backwards-compatible bug fixes.
> 
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

As refinement, and especially to prevent ending up at version 7.2.3 very quickly, adding one additional digit, e.g. 0.x.y.z, similar to the current scheme, could be done.

**Comment by zathras-crypto** [comment]
> Assuming releases are for the broader public, isn't appending -rel sort of just stating the obvious?

Just a simple way for end users to verify they are running a tagged release.  If they are running anything ending in -dev then they are not running a 'release'.

I definitely think we'll do some more work on versioning as we get into a better release process.

**Comment by m21** [comment]
Zathras had changed the versioning paradigm -- close this issue or document it here?
